lib/features/mailbox/domain/usecases/delete_multiple_mailbox_interactor.dart

```diff
@@ -53,4 +78,71 @@ class DeleteMultipleMailboxInteractor {
       yield Left<Failure, Success>(DeleteMultipleMailboxFailure(e));
     }
   }
+
+  MailboxTree buildUnsubscribedMailboxTree(
+    Set<PresentationMailbox> listUnsubscribedMailbox,
+  ) {
+    Map<MailboxId, MailboxNode> mailboxDictionary = HashMap();
+    final unsubscribedTree = MailboxTree(MailboxNode.root());
+
+    for (var mailbox in listUnsubscribedMailbox) {
+      if (mailbox.parentId != null) {
+        mailboxDictionary[mailbox.id] = MailboxNode(mailbox);
+      }
+    }
+
+    for (var mailbox in listUnsubscribedMailbox) {
+      final parentId = mailbox.parentId;
+      if (parentId != null) {
+        final parentNode = mailboxDictionary[parentId];
+        final node = mailboxDictionary[mailbox.id];
+        if (node != null) {
+          if (parentNode != null) {
+            parentNode.addChildNode(node);
+          } else {
+            mailboxDictionary[parentId] = MailboxNode(PresentationMailbox(parentId));
+            unsubscribedTree.root.addChildNode(mailboxDictionary[parentId]!);
+            mailboxDictionary[parentId]!.addChildNode(node);
+          }
+        }
+      }
+    }
+
+    return unsubscribedTree;
+  }
+
+  Map<MailboxId, List<MailboxId>> addUnsubscribedSubFolderToDelete(
+    Map<MailboxId, List<MailboxId>> mapMailboxIdToDelete,
+    List<MailboxId> listMailboxIdToDelete,
+    MailboxTree unsubscribedTree,
+  ) {
+    List<MailboxId> visitedUnsubscribedMailboxIds = [];
+
+    for (var mailboxId in listMailboxIdToDelete) {
+      if (visitedUnsubscribedMailboxIds.contains(mailboxId)) {
+        continue;
+      } else {
+        final matchedNode = unsubscribedTree.findNode((node) => node.item.id == mailboxId);
+
+        if (matchedNode != null) {
+          final descendantIds = matchedNode.descendantsAsList().mailboxIds;
+          final descendantIdsReversed = descendantIds.reversed.toList();
+          descendantIdsReversed.removeLast();
+
+          if (mapMailboxIdToDelete.containsKey(mailboxId)) {
+            mapMailboxIdToDelete[mailboxId]!.insertAll(mapMailboxIdToDelete[mailboxId]!.length - 1, descendantIdsReversed);
+          } else {
+            for (var listIdToDelete in mapMailboxIdToDelete.values) {
```

Comment:
Do we have a faster algorithm? I'm quite worried about using 2 nested loops here.